File: app/src/main/java/com/example/moodlight/screen/main1/CommentAdapter.java
package com.example.moodlight.screen.main1;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.moodlight.R;

import java.util.ArrayList;

public class CommentAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private final int VIEW_TYPE_ITEM = 0;
    private final int VIEW_TYPE_LOADING = 1;

    private ArrayList<String> items;

    public CommentAdapter(ArrayList<String> items) {
        this.items = items;
    }

    @Override
    public int getItemViewType(int position) {
        return items.get(position) == null ? VIEW_TYPE_LOADING :VIEW_TYPE_ITEM;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_ITEM) {
            View vh = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_comment, parent, false);
            return new NormalViewHolder(vh);
        } else {
            View vh = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_loading, parent, false);
            return new LoadingViewHolder(vh);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if(holder instanceof NormalViewHolder){
            onBind((NormalViewHolder) holder, position);
        }
    }

    public void onBind(NormalViewHolder holder, int position){
        String item = items.get(position);
        holder.onBind(item);
    }
    @Override
    public int getItemCount() {
        return items == null ? 0 : items.size();
    }


    private class NormalViewHolder extends RecyclerView.ViewHolder {
        private TextView comment;
        public NormalViewHolder(@NonNull View itemView) {
            super(itemView);
            comment = itemView.findViewById(R.id.comment);
        }
        public void onBind(String item){
            comment.setText(item);
        }
    }
    private class LoadingViewHolder extends RecyclerView.ViewHolder {
        private ProgressBar progressBar;
        public LoadingViewHolder(@NonNull View itemView) {
            super(itemView);
            progressBar = itemView.findViewById(R.id.progressBar);
        }
    }
}

File: app/src/main/java/com/example/moodlight/screen/main1/CommentAdapter.kt
package com.example.moodlight.screen.main1

import android.content.ContentValues.TAG
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.MutableLiveData
import androidx.recyclerview.widget.RecyclerView
import com.example.moodlight.R
import com.example.moodlight.databinding.ItemCommentBinding
import com.example.moodlight.model.CommentModel
import com.example.moodlight.screen.main1.viewmodel.CommentViewModel
import com.example.moodlight.util.DataType

class CommentAdapter() : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var list = ArrayList<CommentModel?>()

    inner class LoadingViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
    }

    inner class NormalViewHolder(private val binding: ItemCommentBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun onBind(item: CommentModel) {
            binding.comment.text = item.contents
            Log.d(TAG, "onBind: ${binding.comment.text}")
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val binding = ItemCommentBinding.inflate(
            LayoutInflater.from(parent.context),comple
            parent,
            false
        )
        val view = LayoutInflater.from(parent.context).inflate(R.layout.row_loading, parent, false)
        return when (viewType) {
            DataType.NORMAL_VIEW_TYPE ->
                NormalViewHolder(binding)
            else ->
                LoadingViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is NormalViewHolder) {
            holder.onBind(list[position]!!)
        }
    }

    override fun getItemCount(): Int = list.size

    fun setItem(list: ArrayList<CommentModel?>) {
        this.list.addAll(list)
        this.list.add(null)
    }

    fun removeAll() {
        this.list.clear()
        notifyDataSetChanged()
    }

    fun removeLoading() {
        if (list.isNotEmpty()) {
            list.removeAt(list.lastIndex)
        }
    }


    override fun getItemViewType(position: Int): Int {
        return if (list[position] == null) {
            DataType.LOADING_VIEW_TYPE
        } else {
            DataType.NORMAL_VIEW_TYPE
        }
    }

}